### PR TITLE
Issue 6904 first free council vote

### PR DESCRIPTION
**Pull request by shamb0**
**Beta implementation to resolve the issue [Participating in Governance should be Free for First Time Voters #6904](https://github.com/paritytech/substrate/issues/6904)**

---

### CODE REVIEW

**Comment by shamb0** [comment]
Hello Shawn,


Like to do quick check on implementation approach for the issue.

In above Demo test sequence, able the detect the first-time voters. But looking for some clarity on balance transaction.
From my initial analysis, looks like fee is not getting detected in all cases ...

**Comment by shamb0** [comment]
Hello Shawn,

Thanks for guidance, reverted pallet-democracy changes.
and modified only pallet-collective.

Used below logic to detect the first vote within a proposal.



And verified with testcase "tests::motions_revoting_works()" [slightly modified] ...

From below trace account 2 is detected as first vote in a proposal ...



**Note:: Plz ignore changes in "frame/democracy/src/tests.rs", Some mistake in my cleanup ... will correct it :-)**

File: frame/collective/src/lib.rs [hunk 1]
```diff
@@ -505,6 +518,7 @@ decl_module! {
 			#[compact] index: ProposalIndex,
 			approve: bool,
 		) -> DispatchResultWithPostInfo {
+			let mut is_account_voting_first_time: bool = false;
```
Comment (shawntabrizi) [inline frame/collective/src/lib.rs:521]:
```suggestion
```

File: frame/collective/src/lib.rs [hunk 2]
```diff
@@ -515,6 +529,22 @@ decl_module! {
 			let position_yes = voting.ayes.iter().position(|a| a == &who);
 			let position_no = voting.nays.iter().position(|a| a == &who);
 
+			// first vote of member in the motion
+			if position_yes.is_none() && position_no.is_none() {
```
Comment (shawntabrizi) [inline frame/collective/src/lib.rs:533]:
```suggestion
			let is_account_voting_first_time = position_yes.is_none() && position_no.is_none();
```

File: frame/collective/src/lib.rs [hunk 3]
```diff
@@ -515,6 +529,22 @@ decl_module! {
...
+				is_account_voting_first_time = true;
+			}
```
Comment (shawntabrizi) [inline frame/collective/src/lib.rs:535]:
```suggestion
			// first vote of member in the motion
			let is_account_voting_first_time = position_yes.is_none() && position_no.is_none();
```

File: frame/collective/src/lib.rs (hunk 1 above)
Comment (shawntabrizi) [reply frame/collective/src/lib.rs:521]: Dont use mutable variables like this if you dont need to

File: frame/collective/src/lib.rs [hunk 4]
```diff
@@ -537,11 +567,29 @@ decl_module! {
...
+						no_votes );
+			}
+
 			Self::deposit_event(RawEvent::Voted(who, proposal, approve, yes_votes, no_votes));
 
 			Voting::<T, I>::insert(&proposal, voting);
 
-			Ok(Some(T::WeightInfo::vote(members.len() as u32)).into())
+			if is_account_voting_first_time {
+				Ok(Pays::No.into())
```
Comment (shawntabrizi) [inline frame/collective/src/lib.rs:588]: You need weight info

File: frame/collective/src/lib.rs (hunk 1 above)
Comment (shamb0) [reply frame/collective/src/lib.rs:521]: Thanks Shawn will correct my bad coding style ...

File: frame/collective/src/lib.rs (hunk 4 above)
Comment (shamb0) [reply frame/collective/src/lib.rs:586]: Fine, Corrected it ...

File: frame/collective/src/lib.rs [hunk 5]
```diff
@@ -537,11 +564,33 @@ decl_module! {
...
+			if is_account_voting_first_time {
+				Ok((Some(
+					T::WeightInfo::vote(members.len() as u32)),
+					Pays::No).into())
+			}
+			else {
+				Ok((Some(
+					T::WeightInfo::vote(members.len() as u32)),
+					Pays::Yes).into())
+			}
```
Comment (shawntabrizi) [inline frame/collective/src/lib.rs:593]:
```suggestion
				Ok((
					Some(T::WeightInfo::vote(members.len() as u32)),
					Pays::No,
				).into())
			}
			else {
				Ok((
					Some(T::WeightInfo::vote(members.len() as u32)),
					Pays::Yes,
				).into())
			}
```

**Comment by shamb0** [comment]
Hello Shawn,

Moved the "mod test" to tests.rs

Introduced new test case "motions_all_first_vote_free_works()" to validate the changes ...

looks like I am messing up in testing for Pay::Yes & Pay::No

For Ok condition I using assert_eq!() pattern



And for Err case I am using match {} pattern



Plz suggest best alternate ...

**Comment by shamb0** [comment]
Plz Ignore this PR, instead use  [ PR | 7661 ](https://github.com/paritytech/substrate/pull/7661)

**Pull request closed without merging**